apis/user/v1alpha1/groupqualityofservicelimits_types.go

```diff
@@ -17,29 +17,86 @@ limitations under the License.
 package v1alpha1
 
 import (
+	"errors"
 	"reflect"
 
+	resource "k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 
 	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
 )
 
+// +kubebuilder:validation:Pattern=`^(0|((0|[1-9][0-9]*)[KMGT]i))$`
+type Quantity string
+
+func (q *Quantity) ToKiB() (*int64, error) {
+	if q == nil {
+		return nil, nil
+	}
+
+	rq, err := resource.ParseQuantity(string(*q))
+	if err != nil {
+		return nil, err
+	}
+
+	i, ok := rq.AsInt64()
+	if !ok {
+		i, ok = rq.AsDec().Unscaled()
+		if !ok {
+			return nil, errors.New("Unable to convert Quantity to KiB int")
+		}
+	}
+
+	i /= 1024
+	return &i, nil
+}
+
+// QualityOfService configures data limits. The value -1 indicates unlimited.
+type QualityOfServiceLimits struct {
+	// StorageQuotaBytes is the limit for total stored data in bytes.
+	//+optional
+	StorageQuotaBytes *Quantity `json:"storageQuotaBytes,omitempty"`
+	// StorageQuotaCount is the limit for total number of objects.
+	//+optional
+	StorageQuotaCount *uint32 `json:"storageQuotaCount,omitempty"`
+	// RequestsPerMin is the limit for number of HTTP requests per minute.
+	//+optional
+	RequestsPerMin *uint32 `json:"requestsPerMin,omitempty"`
+	// InboundBytesPerMin is the limit for inbound data per minute in bytes.
+	//+optional
+	InboundBytesPerMin *Quantity `json:"inboundBytesPerMin,omitempty"`
+	// OutboundKiBsPerMin is the limit for outbound data per minute in bytes.
+	//+optional
+	OutboundBytesPerMin *Quantity `json:"outboundBytesPerMin,omitempty"`
+}
```

Comment:
Input from MJ in todays SUP: Can we make the fields of this struct nullable to enable users to explicitly set an "unlimited" quota?